library/Backends/Michelson/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs

```diff
@@ -423,71 +447,124 @@ onPairGen1 op f =
          in Env.Constant (f (car, cdr))
     )
 
-onTwoArgs :: OnTerm2 m (V.Value' Types.Op) Env.Expanded
-onTwoArgs op f typ instrs = do
+pushConstant :: Env.Reduction m => Types.Type -> [Types.RawTerm] -> m Env.Expanded
+pushConstant =
+  onOneArgGen
+    ( \instr1 -> do
+        addExpanded instr1
+        case val instr1 of
+          Env.Constant _ -> pure Env.Nop
+          _ -> pure (val instr1)
+    )
+
+--------------------------------------------------------------------------------
+-- On Argument Functions
+--------------------------------------------------------------------------------
+
+------------------------------------------------------------
+-- Generator functions
+------------------------------------------------------------
+
+onTwoArgsGen :: OnTerm2Gen m Protect Env.Expanded
+onTwoArgsGen applyOperation typ arguments = do
   -- last argument evaled first
-  v <- traverse (protect . (inst >=> promoteTopStack)) (reverse instrs)
+  v <- traverse (protect . (inst >=> promoteTopStack)) (reverse arguments)
```

Comment:
we could abstract this, there's code in common between `onTwoArgsGen` and `onOneArgGen`

Reply:
yeah I could probably with this, didn't think of it before